In [1]:
import pandas as pd

In [2]:
data_dir = "./data/raw_data/"
save_dir = "./data/preprocessed_data/updated_data/"

Read in the data

In [3]:
cv_table = pd.read_csv(data_dir + "updated_data/CRR.csv")
ndr_table = pd.read_csv(data_dir + "updated_data/WQR.csv")
poll_table = pd.read_csv(data_dir + "updated_data/POLL.csv")

latlong_cv = pd.read_csv(data_dir + "latlong_cv.csv")
latlong_ndr = pd.read_csv(data_dir + "latlong_ndr.csv")
latlong_pollination = pd.read_csv(data_dir + "latlong_pollination.csv")


We have three tables of data, one for each contribution of nature. We also have three tables relating each data point to a longitude and a latitude.

Let's look at the tables first!

# cv_table

In [4]:
cv_table.describe(include = 'all')

fid    country         region           UN_c           UN_1  \
count   686665.000000     685971         685971  686665.000000  686665.000000   
unique            NaN        161              8            NaN            NaN   
top               NaN  Indonesia  North America            NaN            NaN   
freq              NaN      74263         160461            NaN            NaN   
mean    343464.696723        NaN            NaN       2.483391       2.823775   
std     198006.613779        NaN            NaN       0.511871       0.573231   
min       3304.000000        NaN            NaN       1.030057       1.237033   
25%     171667.000000        NaN            NaN       2.094444       2.376177   
50%     343333.000000        NaN            NaN       2.449490       2.786279   
75%     514999.000000        NaN            NaN       2.826065       3.259844   
max     686665.000000        NaN            NaN       4.291871       4.817462   

                 UN_3           UN_5           NC_c           NC_1  \
count   686665.000000  686665.000000  686665.000000  686665.000000   
unique            NaN            NaN            NaN            NaN   
top               NaN            NaN            NaN            NaN   
freq              NaN            NaN            NaN            NaN   
mean         2.984508       3.140089       0.037279       0.036749   
std          0.617884       0.644807       0.078843       0.078211   
min          1.297790       1.311983      -1.000000      -1.000000   
25%          2.492883       2.667168       0.000000       0.000000   
50%          2.935599       3.107233       0.000000       0.000000   
75%          3.419952       3.619904       0.092263       0.066752   
max          5.000000       5.000000       0.235276       0.235276   

                 NC_3           NC_5         pop_c         pop_1  \
count   686665.000000  686665.000000  6.866650e+05  6.866650e+05   
unique            NaN            NaN           NaN           NaN   
top               NaN            NaN           NaN           NaN   
freq              NaN            NaN           NaN           NaN   
mean         0.036893       0.036704  1.126055e+03  1.373949e+03   
std          0.078365       0.078179  7.320693e+04  9.411554e+04   
min         -1.000000      -1.000000  0.000000e+00  0.000000e+00   
25%          0.000000       0.000000  4.970154e-01  4.151525e-01   
50%          0.000000       0.000000  2.177214e+01  2.147821e+01   
75%          0.071753       0.062181  1.606435e+02  1.571138e+02   
max          0.235276       0.235276  3.145269e+07  4.268553e+07   

               pop_3         pop_5  
count   6.866650e+05  6.866650e+05  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean    1.421939e+03  1.398819e+03  
std     9.292145e+04  9.330428e+04  
min     0.000000e+00  0.000000e+00  
25%     4.487666e-01  4.361424e-01  
50%     2.678465e+01  2.193834e+01  
75%     1.878266e+02  1.628649e+02  
max     4.358711e+07  4.143670e+07

This table describes the coastal risk mitigation by natural habitats.
Some explanation of the table follow:

* **fid** - the id of the geographical location that the data describes. This id can be converted to lon/lat using the latlong table.  
* **UN_cur** - the **current** (2015) unmet need of the coastal risk mitigation in this particular area. If this figure is negative then it means that we have enough, and even a surplus, in this particular area. 
* **UN_ssp1** the unmet need of the coastal risk mitigation in the year 2050, given that scenario **ssp1** is realized. 
* **ssp1, ssp3, ssp5** - the three future scenarios. It is not perfectly clear which scenraio is which here. ssp1 is most likely the sustainability scenario, purely based on the observation that the unmet need in this column is lower than for the other scenarios. The other two scenarios are fossil-fuels and regional rivalry.
* **pop_cur** - the **current** (2015) number of people in this geographic region who are affected by coastal risk mitigation. If the unmet need is high and the population is high then a lot of people are negatively affected.
* **pop_sspX** - same as the above, but in the year 2050, for the three different future scenarios.

# ndr_table

In [5]:
ndr_table.describe(include = 'all')

fid country  region          UN_c          UN_1  \
count   13215.000000   13215   13215  1.321500e+04  1.321500e+04   
unique           NaN     168       7           NaN           NaN   
top              NaN  Russia  Africa           NaN           NaN   
freq             NaN    1506    2738           NaN           NaN   
mean    24756.199924     NaN     NaN  5.439896e+06  5.696458e+06   
std     10348.534721     NaN     NaN  1.053685e+07  1.291153e+07   
min      9925.000000     NaN     NaN  0.000000e+00  0.000000e+00   
25%     15790.500000     NaN     NaN  7.382453e+05  4.839964e+05   
50%     22568.000000     NaN     NaN  2.256096e+06  1.603293e+06   
75%     32980.500000     NaN     NaN  5.625936e+06  5.526796e+06   
max     52313.000000     NaN     NaN  1.714816e+08  3.159408e+08   

                UN_3          UN_5          NC_c          NC_1          NC_3  \
count   1.321500e+04  1.321500e+04  13215.000000  13215.000000  13215.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean    6.617717e+06  5.517637e+06      0.722226      0.758806      0.712588   
std     1.410886e+07  1.121654e+07      0.155923      0.145997      0.230949   
min     0.000000e+00  0.000000e+00      0.268032      0.133294     -2.215372   
25%     5.232498e+05  4.904952e+05      0.616834      0.670287      0.612981   
50%     1.879288e+06  1.738118e+06      0.739356      0.793527      0.773301   
75%     6.390241e+06  5.948842e+06      0.845648      0.855730      0.857030   
max     2.093042e+08  2.064157e+08      1.000000      1.000000      1.000000   

                NC_5         pop_c         pop_1         pop_3         pop_5  
count   13215.000000  1.321500e+04  1.183900e+04  1.183900e+04  1.183900e+04  
unique           NaN           NaN           NaN           NaN           NaN  
top              NaN           NaN           NaN           NaN           NaN  
freq             NaN           NaN           NaN           NaN           NaN  
mean        0.741928  3.233945e+05  2.150300e+05  4.964972e+05  2.140241e+05  
std         0.217848  9.299560e+05  5.734855e+05  1.387242e+06  5.675742e+05  
min        -3.048947  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%         0.661652  3.529842e+03  5.168572e+03  7.882555e+03  5.465063e+03  
50%         0.794158  3.726848e+04  3.035928e+04  5.353800e+04  3.156906e+04  
75%         0.863073  2.124498e+05  1.431630e+05  3.170438e+05  1.455778e+05  
max         1.000000  1.667823e+07  8.933015e+06  2.162865e+07  8.607843e+06

This table describes the improvement of water quality by filtering of pollutants (in our case probably excessive amounts of nutrients, making the water quality bad) when water passes by natural habitats such as forests and wetlands.

The columns are the same as in the previous table, although `pop` is replaced by `rurpop`. Charlotte has confirmed that this is still the same metric though.

We also have two extra columns `country` and `region`. We will most probably not use these, especially as we don't actually have them for each table.

One thing we can immediately observe is that the count (number of entries) is much lower, the previous table had 686665 entries while this one has only 25243. This needs to be investigated. Does this mean that we don't have data for the whole earth for this particular contribution of nature?

# poll_table

In [6]:
poll_table.describe(include = 'all')

fid country   region          UN_c          UN_1  \
count    9160.000000    9160     9160  9.160000e+03  9.160000e+03   
unique           NaN     161        7           NaN           NaN   
top              NaN  Russia  Eurasia           NaN           NaN   
freq             NaN     876     1951           NaN           NaN   
mean    25270.656441     NaN      NaN  9.630618e+04  8.972863e+04   
std     10032.053267     NaN      NaN  4.348216e+05  4.136170e+05   
min      8845.000000     NaN      NaN  0.000000e+00  0.000000e+00   
25%     16672.000000     NaN      NaN  2.088596e+00  2.046068e+00   
50%     22971.500000     NaN      NaN  1.027879e+03  9.781146e+02   
75%     33446.250000     NaN      NaN  3.107308e+04  2.827017e+04   
max     51951.000000     NaN      NaN  2.322062e+07  2.249534e+07   

                UN_3          UN_5         NC_c         NC_1         NC_3  \
count   9.160000e+03  9.160000e+03  9160.000000  9160.000000  9160.000000   
unique           NaN           NaN          NaN          NaN          NaN   
top              NaN           NaN          NaN          NaN          NaN   
freq             NaN           NaN          NaN          NaN          NaN   
mean    9.375806e+04  9.990251e+04     0.342024     0.363090     0.291586   
std     4.113846e+05  4.335244e+05     0.368788     0.371759     0.354174   
min     0.000000e+00  0.000000e+00     0.000000     0.000000     0.000000   
25%     9.315915e+00  9.550000e+00     0.017078     0.019841     0.007737   
50%     1.492830e+03  1.594218e+03     0.161952     0.204912     0.089837   
75%     3.439839e+04  3.474784e+04     0.666667     0.685986     0.552810   
max     2.321814e+07  2.270223e+07     1.000000     1.000000     1.000000   

               NC_5         pop_c         pop_1         pop_3         pop_5  
count   9160.000000  9.160000e+03  6.540000e+03  6.540000e+03  6.540000e+03  
unique          NaN           NaN           NaN           NaN           NaN  
top             NaN           NaN           NaN           NaN           NaN  
freq            NaN           NaN           NaN           NaN           NaN  
mean       0.312048  4.576569e+05  9.023251e+05  1.093869e+06  8.753412e+05  
std        0.352050  1.481439e+06  2.415334e+06  2.722593e+06  2.405526e+06  
min        0.000000  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
25%        0.014991  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  
50%        0.134070  3.016314e+04  1.452440e+05  1.897702e+05  1.260224e+05  
75%        0.592607  2.764953e+05  6.747889e+05  8.782078e+05  6.261508e+05  
max        1.000000  3.046985e+07  3.778232e+07  3.752056e+07  3.774713e+07

This table describes the pollination of crops coming from natural habitats inhabiting pollinators (ex. bees).

The columns are the same as previously.

Here though we have an even lower count. What does this mean? A hypothesis could be that the resolution is simply not just not as high for the data here.

# latlong_cv

In [7]:
latlong_cv.describe(include = 'all')

fid            lat            lng
count  686665.000000  686665.000000  686665.000000
mean   343333.000000       9.346826      23.706301
std    198223.255633      32.202816     102.750279
min         1.000000     -58.480000    -180.000000
25%    171667.000000     -10.600000     -73.070000
50%    343333.000000       9.600000      12.640000
75%    514999.000000      35.450000     123.250000
max    686665.000000      62.800000     180.000000

This looks good, just a simple mapping from the **fid** to the coordinates. We have the same number of mappings as we have entrys in the `cv_table`.

# latlong_ndr

In [8]:
latlong_ndr.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Here the number of entries is actually higher than the number of entries in the `ndr_table`. This is a bit strange, what does this mean?

Moreover, why do we even need to have different mappings for the different contributions of nature? Couldn't they just use the same mappings between id and coordinates? The answer to this is probably that, as suggested before, the world map resolutions of the nature's contributions are different, with coastal risk having the most detailed data.

# latlong_pollination

In [9]:
latlong_pollination.describe(include = 'all')

fid           lat           lng
count  64800.000000  64800.000000  64800.000000
mean   32399.500000      0.000000      0.000000
std    18706.293059     51.961123    103.923449
min        0.000000    -89.500000   -179.500000
25%    16199.750000    -44.750000    -89.750000
50%    32399.500000      0.000000      0.000000
75%    48599.250000     44.750000     89.750000
max    64799.000000     89.500000    179.500000

Once again the same thing, the number of entries are the same as in the `latlong_ndr` table, 64800, while the number of entries in the `poll_table` is just 9160...



# Actual data

We also take a look at the actual data, just to get a quick feel for the actual figures:

In [10]:
cv_table.head()

fid country   region      UN_c      UN_1      UN_3      UN_5  NC_c  NC_1  \
0  9715  Russia  Eurasia  2.591935  2.909349  3.112751  3.265634  -1.0  -1.0   
1  9716  Russia  Eurasia  2.865853  3.216811  3.441708  3.610748  -1.0  -1.0   
2  9717  Russia  Eurasia  2.993380  3.359955  3.594861  3.914325  -1.0  -1.0   
3  9718  Russia  Eurasia  2.720859  3.054061  3.267581  3.557960  -1.0  -1.0   
4  9719  Russia  Eurasia  2.947428  3.308376  3.539675  3.854235  -1.0  -1.0   

   NC_3  NC_5      pop_c      pop_1      pop_3      pop_5  
0  -1.0  -1.0   2.137938   2.137938   2.137938   2.137938  
1  -1.0  -1.0  11.744734  11.744734  11.744734  11.744734  
2  -1.0  -1.0  39.133438  39.133438  39.133438  39.133438  
3  -1.0  -1.0  30.200581  30.200581  30.200581  30.200581  
4  -1.0  -1.0   2.611755   2.611755   2.611755   2.611755

Nothing surprising here! The id:s start from 1 and go up. We see an unmet need for the first 5 geographical regions both currently and for the 3 different future scenarios. We also see that nobody in this region actually gets directly affected by any changes to the natural habitats mitigating coastal risk. This could be because nobody lives there, or simply perhaps because there is no coast in this area.

Let's look at the `ndr_table`.

In [11]:
ndr_table.sample(5)

fid   country         region          UN_c         UN_1          UN_3  \
4251  17542     China     North Asia  2.084198e+07  15454076.00  1.578908e+07   
8626  28266   Nigeria         Africa  4.005714e+06   7356395.50  5.402769e+06   
9860  32878    Brazil  South America  1.014693e+06   1057754.50  1.029648e+06   
8479  27638     Burma     South Asia  4.659542e+05    602906.25  5.129200e+05   
3065  15402  Mongolia     North Asia  3.807993e+06   3956525.00  3.964273e+06   

              UN_5      NC_c      NC_1      NC_3      NC_5         pop_c  \
4251  1.612741e+07  0.401888  0.527434  0.546894  0.537184  6.696001e+05   
8626  1.362406e+07  0.556918  0.558537  0.402386 -0.506992  1.182646e+06   
9860  1.034690e+06  0.843875  0.841125  0.841574  0.840798  5.905507e+03   
8479  5.341347e+05  0.872516  0.853717  0.859666  0.853862  2.684109e+05   
3065  4.082686e+06  0.519492  0.505114  0.499772  0.484830  0.000000e+00   

             pop_1         pop_3         pop_5  
4251  5.408986e+05  6.315461e+05  5.446566e+05  
8626  1.499393e+06  3.492779e+06  1.494299e+06  
9860  6.104537e+03  6.978535e+03  6.107604e+03  
8479  2.255478e+05  3.009044e+05  2.165765e+05  
3065           NaN           NaN           NaN

We note that the **fid** goes above the maximum number of entries in the table (which was 25243). 

Same for `poll_table`:

In [12]:
poll_table.sample(5)

fid        country         region           UN_c           UN_1  \
4206  21539          China     North Asia  303752.862689  238437.818204   
8319  40449         Brazil  South America   87246.510502   86276.349220   
3845  20411       Pakistan     South Asia  151823.115818  150478.402759   
4996  25062   Saudi Arabia        Eurasia     228.483547     355.905349   
1930  15912  United States  North America       2.880028       2.218082   

               UN_3           UN_5      NC_c      NC_1          NC_3  \
4206  210546.758871  334684.507434  0.117193  0.188207  1.702534e-01   
8319  159030.300940  164307.880266  0.504894  0.509686  4.337873e-01   
3845  240485.321148  238175.902047  0.257521  0.252546  1.266498e-01   
4996    1114.530038      99.344533  0.578756  0.142170  3.027551e-17   
1930       6.887695       6.144579  0.067950  0.143052  3.974774e-02   

          NC_5         pop_c        pop_1      pop_3       pop_5  
4206  0.156373  0.000000e+00        0.000        0.0        0.00  
8319  0.475474  0.000000e+00          NaN        NaN         NaN  
3845  0.132689  2.327554e+06  3312313.750  4972233.0  3227698.25  
4996  0.593631  1.605191e+05   294045.875   332083.5   320683.00  
1930  0.107386  0.000000e+00        0.000        0.0        0.00

And for an example lat_long-table let's check out `latlong_cv`:

In [13]:
latlong_cv.sample(5)

fid    lat     lng
132293  132294 -16.60   49.80
491228  491229  30.61  121.14
288274  288275   0.44  120.99
77875    77876 -37.07  175.88
621794  621795  53.93   -2.98

# Preprocessing

The next step is to perform the preprocessing (if needed) and save the data in new files. 
This might need to be done in several iteration as we don't know at the moment exactly how we will use the data. 

We first replace any NaN-values with zeroes:

## Replace NaN with zeroes:

The tables have NaN:s in them from reading empty entriees in the csv files:

In [14]:
poll_table.isna().values.any()

True

We replace these with 0 (the fid columns are never NaN).

In [15]:
cv_table.fillna(value=0, inplace=True)
ndr_table.fillna(value=0, inplace=True)
poll_table.fillna(value=0, inplace=True)

Another thing that will be good to have done already is to join the lat/lon tables with the data tables.



## Join lat/lon tables with the data tables

Start with `cv_table`:

In [16]:
cv_table_out = pd.merge(cv_table, latlong_cv, on='fid')
cv_table_out.sample()

fid    country      region      UN_c      UN_1     UN_3      UN_5  \
188395  188396  Indonesia  South Asia  1.944161  2.334815  2.44949  2.542303   

            NC_c      NC_1      NC_3      NC_5      pop_c      pop_1  \
188395  0.156567  0.156567  0.156567  0.156567  113.13201  88.962899   

             pop_3      pop_5  lat     lng  
188395  137.815358  88.448663 -7.3  120.87

`fid`, `country` or `region` column is not needed, drop this.

In [17]:
cv_table_out.drop(columns=['fid', 'country', 'region'], inplace=True)
cv_table_out.sample()

UN_c      UN_1      UN_3     UN_5  NC_c  NC_1  NC_3  NC_5  \
25266  3.383363  3.797696  3.797696  4.26277   0.0   0.0   0.0   0.0   

          pop_c     pop_1     pop_3     pop_5    lat    lng  
25266  0.426638  0.426638  0.476831  0.451734 -52.26 -74.38

Looks good! Let's do the same for `ndr_table` and `poll_table`:

In [18]:
ndr_table_out = pd.merge(ndr_table, latlong_ndr, on ='fid')

# Remove the country and region columns, which we will probably not use
ndr_table_out.drop(columns=['fid', 'country', 'region'], inplace=True) 

ndr_table_out.sample()

UN_c         UN_1          UN_3         UN_5      NC_c  \
2509  491196.84375  501382.9375  512116.78125  503447.0625  0.864589   

          NC_1      NC_3      NC_5       pop_c       pop_1       pop_3  \
2509  0.864346  0.858822  0.861212  443.757507  536.473755  343.222534   

           pop_5   lat   lng  
2509  723.209656  49.5 -72.5

In [19]:

poll_table_out = pd.merge(poll_table, latlong_pollination, on='fid')
poll_table_out.drop(columns=['fid', 'country', 'region'], inplace=True)

poll_table_out.sample()

UN_c  UN_1      UN_3     UN_5      NC_c  NC_1      NC_3      NC_5  \
3530  0.214282   0.0  0.261955  0.21227  0.999264   1.0  0.999071  0.999216   

            pop_c        pop_1       pop_3        pop_5   lat   lng  
3530  8682.703125  7641.498535  5807.07959  9261.086914  35.5 -75.5

## Replace NaN with zeroes

The ndr and poll tables have NaN-values (from reading in

## Dropping rows without any inhabitants
We are interested in the number of affected people. Therefore, we can drop the data rows where there the population is 0 (currently, and for all of the 3 scenarios in 2050), e.g. places which are inhabited.

In [20]:
cols = ['pop_c', 'pop_1', 'pop_3', 'pop_5']
print("Number of rows before: " + str(cv_table_out.shape[0]))
cv_table_out = cv_table_out.loc[(cv_table_out[cols]!=0).any(axis=1)]
print("Number of rows after: " + str(cv_table_out.shape[0]))

Number of rows before: 686665
Number of rows after: 635204


Doing the same for the ndr_table and the poll table:

In [21]:
print("Number of rows in ndr_table before: " + str(ndr_table_out.shape[0]))
ndr_table_out = ndr_table_out.loc[(ndr_table_out[cols]!=0).any(axis=1)]
print("Number of rows in ndr_table after: " + str(ndr_table_out.shape[0]))

print("")

print("Number of rows in poll_table before: " + str(poll_table_out.shape[0]))
poll_table_out = poll_table_out.loc[(poll_table_out[cols]!=0).any(axis=1)]
print("Number of rows in poll_table after: " + str(poll_table_out.shape[0]))

Number of rows in ndr_table before: 13215
Number of rows in ndr_table after: 11913

Number of rows in poll_table before: 9160
Number of rows in poll_table after: 6487


A lot of data is lost doing this, especially proportionately in the case of the ndr and the poll datasets.

This is fine though since, as stated before, we are only interested in the number of affected people.

*In this iteration 2, we note that the ndr_table was already cleaned of empty population rows. This strenghtens the choice of removing them, and we continue doing so in this updated data preprocessing*.

## Subsampling data

The cv dataset is a bit too large to be used as it is (will have too long loading times).
Therefore we subsample it to be smaller:

In [22]:
cv_table_out = cv_table_out.sample(30000)

Now we are ready to save the files!

# Saving data

Doing a final reset of the indexes:

In [23]:
cv_table_out.reset_index(drop=True, inplace=True)
ndr_table_out.reset_index(drop=True, inplace=True)
poll_table_out.reset_index(drop=True, inplace=True)

In [24]:
cv_table_out.to_csv(save_dir + "cv_table_preprocessed.csv", index=False)
ndr_table_out.to_csv(save_dir + "ndr_table_preprocessed.csv", index=False)
poll_table_out.to_csv(save_dir + "poll_table_preprocessed.csv", index=False)